### PR TITLE
Add shutdown argument to VBCSCompiler

### LINKED ISSUE #7115: vbcscompiler.exe should support a /shutdown flag
The vbcscompiler.exe should have some nice way to shut down itself down without resorting to killing all instances everywhere. I could imagine it has a /shutdown flag that either:

1. Shuts down running instances that match the invoked vbcscompiler.exe by full path, or
2. Takes a path and shuts down running instances from that path.

### PR DESCRIPTION
**Pull request by jaredpar**
This adds a `-shutdown` argument to VBCSCompiler that allows clients to gracefully shutdown the server.  This is implemented by adding a new message to the protocol which does two actions:

- Begins a shut down of the server.
- Returns the process id of the server to the client who initiated the shutdown.  

When reviewing the changes to `ServerDispatcher` it's important to understand the difference between the following operations:

- Shutting down: The server will accept no new build requests but will allow previously scheduled ones to complete without interruption.  While these builds are finishing up the server will continue to respond to shutdown requests with the process id. 
- Cancelling: This happens when the `CancellationToken` passed into the server is signaled.  In this case the server will cancel all previously scheduled build requests and accept no new connections.  In other words, cancel as fast as possible. 

closes #7115

CC @dotnet/roslyn-compiler

### CODE REVIEW

**Comment by jaredpar** [comment]
retest please

**Comment by davkean** [comment]
@dotnet-bot retest prtest/lin/dbg/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_lin_dbg_unit32/2713/
// Retest reason: Build break

**Comment by agocke** [comment]
What's the motivation here?

**Comment by jaredpar** [comment]
@agocke there were an increasing number of scenarios where we wanted to shutdown a specific compiler server gracefully without waiting for the timeout to expire.  Typically in order to get it to release references to the compiler DLLs.  The scenarios we had hit:

- The Jenkins build scripts
- NuGet operations
- Roslyn VSIX install / uninstall 

Killing all VBCSCompiler instances, while safe, is a bit hacky.  Having a definitive shutdown seemed like a much more palatable option.

File: src/Compilers/Core/CommandLine/BuildProtocol.cs [hunk 1]
```diff
@@ -268,7 +256,9 @@ public enum ResponseType
         {
             MismatchedVersion,
             Completed,
-            AnalyzerInconsistency
+            AnalyzerInconsistency,
+            Shutdown,
```
Comment (AlekseyTs) [inline src/Compilers/Core/CommandLine/BuildProtocol.cs]: What does response Shutdown mean?

Comment (AlekseyTs) [inline src/Compilers/Core/CommandLine/BuildProtocol.cs]: I think members of this enum deserve good comments.

File: src/Compilers/Server/ServerShared/ServerDispatcher.cs [hunk 2]
```diff
@@ -166,32 +190,116 @@ private void WaitForAnyCompletion(IEnumerable<Task<ConnectionData>> e, Task[] ot
...
+            _state = State.ShuttingDown;
+        }
+
+        private void HandleCompletedGCTask()
+        {
+            _gcTask = null;
+            GC.Collect();
+        }
+
+        private void MaybeCreateTimeoutTask()
```
Comment (AlekseyTs) [inline src/Compilers/Server/ServerShared/ServerDispatcher.cs]: "Maybe" in the name feels strange. How about changing name to CreateTimeoutTaskIfNeeded? Similar suggestion for other similar names.

**Comment by AlekseyTs** [comment]
LGTM

**Comment by VSadov** [comment]
LGTM

**Pull request merged**